src/Ide/Plugin/Tactic.hs

```diff
@@ -0,0 +1,281 @@
+-- | Terminal constructor for providing context-sensitive tactics. Tactics
+-- given by 'provide' are always available.
+provide :: TacticCommand -> T.Text -> T.Text -> TacticProvider
+provide tc title name plId uri range _ = do
+  let params = TacticParams { file = uri , range = range , var_name = name }
+  cmd <- mkLspCommand plId (tcCommandId tc) title (Just [toJSON params])
+  pure
+    $ pure
+    $ CACodeAction
+    $ CodeAction title (Just CodeActionQuickFix) Nothing Nothing
+    $ Just cmd
+
+
+------------------------------------------------------------------------------
+-- | Restrict a 'TacticProvider', making sure it appears only when the given
+-- predicate holds for the goal.
+filterGoalType :: (Type -> Bool) -> TacticProvider -> TacticProvider
+filterGoalType p tp plId uri range jdg@(Judgement _ (CType g)) =
+  case p g of
+    True  -> tp plId uri range jdg
+    False -> pure []
+
+
+------------------------------------------------------------------------------
+-- | Multiply a 'TacticProvider' for each binding, making sure it appears only
+-- when the given predicate holds over the goal and binding types.
+filterBindingType
+    :: (Type -> Type -> Bool)  -- ^ Goal and then binding types.
+    -> (OccName -> Type -> TacticProvider)
+    -> TacticProvider
+filterBindingType p tp plId uri range jdg@(Judgement hys (CType g)) =
+  fmap join $ for (M.toList hys) $ \(occ, CType ty) ->
+    case p g ty of
+      True  -> tp occ ty plId uri range jdg
+      False -> pure []
+
+
+data TacticParams = TacticParams
+    { file :: Uri -- ^ Uri of the file to fill the hole in
+    , range :: Range -- ^ The range of the hole
+    , var_name :: T.Text
+    }
+  deriving (Show, Eq, Generic, ToJSON, FromJSON)
+
+
+------------------------------------------------------------------------------
+-- | Find the last typechecked module, and find the most specific span, as well
+-- as the judgement at the given range.
+judgmentForHole
+    :: IdeState
+    -> NormalizedFilePath
+    -> Range
+    -> IO (Maybe (PositionMapping, LHsExpr GhcTc, Judgement))
+judgmentForHole state nfp range = runMaybeT $ do
+  (tmr, pos) <- MaybeT $ runIde state $ useWithStale TypeCheck nfp
+  range' <- liftMaybe $ fromCurrentRange pos range
+  let span = rangeToSrcSpan (fromNormalizedFilePath nfp) range'
+      mod = tmrModule tmr
+
+  (mss@(L span' (HsVar _ (L _ v))))
+    <- liftMaybe $ mostSpecificSpan @_ @GhcTc span (tm_typechecked_source mod)
+
+  let goal = varType v
+      binds = bindings mod
+      hyps = hypothesisFromBindings span' binds
+  pure (pos, holify binds mss, Judgement hyps $ CType goal)
+
+
+tacticCmd :: (OccName -> TacticsM ()) -> CommandFunction TacticParams
+tacticCmd tac lf state (TacticParams uri range var_name)
+  | Just nfp <- uriToNormalizedFilePath $ toNormalizedUri uri =
+      fromMaybeT (Right Null, Nothing) $ do
+        (pos, _, jdg) <- MaybeT $ judgmentForHole state nfp range
+        let dflags = unsafeGlobalDynFlags
```

Comment:
To get the actual dflags for the module use the `GhcSession` rule

Reply:
Did it via `GetModIface`, but should be same-same, right?